File: clients/ios/StellarChat/StellarChat/Models/ContactAliasStore.swift
import Foundation

/// In-memory cache of contact aliases (pubkey → display name), backed by
/// encrypted SwiftData persistence. All views observe `aliases` reactively.
@MainActor
@Observable
final class ContactAliasStore {
    private(set) var aliases: [String: String] = [:]
    private let store: PersistenceStore

    init(store: PersistenceStore) {
        self.store = store
        self.aliases = store.loadContactAliases()
    }

    /// Returns the display name for a pubkey, or nil if no alias is set.
    func displayName(for pubkey: String) -> String? {
        aliases[pubkey]
    }

    /// Set or update an alias for a pubkey. Persists encrypted to disk.
    func setAlias(pubkey: String, name: String) {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            removeAlias(pubkey: pubkey)
            return
        }
        aliases[pubkey] = trimmed
        store.saveContactAliasAsync(pubkey: pubkey, name: trimmed)
    }

    /// Remove an alias for a pubkey.
    func removeAlias(pubkey: String) {
        aliases.removeValue(forKey: pubkey)
        store.deleteContactAliasAsync(pubkey: pubkey)
    }
}

File: clients/android/StellarChat/app/src/main/java/chat/onym/android/persistence/ContactAliasStore.kt
package chat.onym.android.persistence

import androidx.compose.runtime.mutableStateMapOf
import chat.onym.android.crypto.StorageEncryption

/**
 * In-memory cache of contact aliases (pubkey → display name), backed by
 * encrypted Room persistence. Compose State map triggers recomposition on changes.
 */
class ContactAliasStore {
    val aliases = mutableStateMapOf<String, String>()

    suspend fun load(dao: StellarChatDao) {
        val persisted = dao.loadAllAliases()
        for (alias in persisted) {
            try {
                val name = StorageEncryption.decryptString(alias.encryptedName)
                aliases[alias.pubkey] = name
            } catch (_: Exception) { /* skip corrupted entries */ }
        }
    }

    fun displayName(pubkey: String): String? = aliases[pubkey]

    suspend fun setAlias(pubkey: String, name: String, dao: StellarChatDao) {
        val trimmed = name.trim()
        if (trimmed.isEmpty()) {
            removeAlias(pubkey, dao)
            return
        }
        aliases[pubkey] = trimmed
        dao.saveContactAlias(
            PersistedContactAlias(
                pubkey = pubkey,
                encryptedName = StorageEncryption.encrypt(trimmed),
                updatedAt = System.currentTimeMillis()
            )
        )
    }

    suspend fun removeAlias(pubkey: String, dao: StellarChatDao) {
        aliases.remove(pubkey)
        dao.deleteContactAlias(pubkey)
    }
}
